### PR TITLE
RFD 89: Merge Webapps into Teleport

### PR DESCRIPTION
**Pull request by hatched**
The RFD for the outline of merging the webapps and webapps.e repositories into the teleport and teleport.e repositories. 

Updates #15164

### CODE REVIEW

File: rfd/0089-merge-webapps.md
Comment (zmb3) [inline rfd/0089-merge-webapps.md:24]: Might be worth adding that the webapps monorepo was developed at a time when we were supporting the UI for both Gravity and Teleport. Now that Gravity is no more, it makes more sense to combine them.

File: rfd/0089-merge-webapps.md
Comment (zmb3) [inline rfd/0089-merge-webapps.md:48]:
```suggestion
  - [ ] Merge remaining PRs.
```

File: rfd/0089-merge-webapps.md
Comment (zmb3) [inline rfd/0089-merge-webapps.md:102]: Would this job go away?

**Review by zmb3 — COMMENTED** [review]
Let's add a few more sections:

1. Developer experience. 
  * summarize today's workflow for a fullstack dev testing local changes. how does it change after this migration?
  * summarize the amount of steps that have to be taken to backport a change (in the worst case, it's merge PR to webapps.e, then merge PR to webapps with updated e ref, then wait for webassets to be updated by GHA, then merge PR to teleport that updates webassets,..)
2. Build process w.r.t webassets. I assume we'll be removing webassets and Teleport will no longer contain a compiled version of the webapp. This means people will need node/yarn (and we should also offer a dockerized version)
  * We also have a GitHub service account and a GitHub action that automatically updates webassets. This will go away and we can deprovision the service account. (cc @adaadb6)
3. Teleport Connect
  * The release process currently requires that we create the same tag on webapps as we do for teleport in order to clone the right version of webapps. This will go away.
  * There's a number of `// TODO` in dronegen that will also be removed since the teleport connect code will already be present by default
  * Proto generation (see #15955 and work with @ravicious on this) 
4. Git history. Is it possible to maintain this? https://stackoverflow.com/questions/13040958/merge-two-git-repositories-without-breaking-file-history

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:102]: Yes it would! Updated.

We should ensure that it can still be built though so we should create a new job to tackle that.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:24]: done

**Comment by zmb3** [comment]
Another thing to consider is backports. If we put this in place in master after we cut branch/v11, then that means we'll be committing most webapps changes to teleport/master from that point on, but backports to v11, v10, and v9 will require updates to webapps (unless we decide to backport the migration to those branches as well). What would be the pros/cons of this approach?

File: rfd/0089-merge-webapps.md
Comment (ravicious) [inline rfd/0089-merge-webapps.md:144]: Isn't ensuring that webassets can be built the thing that this job does already?

Also, the `push-build-darwin-amd64` pipeline already runs `yarn install`, `yarn build-term` and `yarn package-term`. However, this only happens after a commit lands in teleport's master. I don't know if we do it for other platforms as well, we'd need to look into the dronegen stuff. The question is whether we want the push pipelines to still run on every commit to teleport.

The reason Connect is build during `push-build-darwin-amd64` is that it's one of the artifacts published on tag release. If there are problems with building Connect, [we want to discover them ASAP rather than wait for a tag release to break](https://gravitational.slack.com/archives/C0DF0TPMY/p1654618901893779?thread_ts=1654598627.161199&cid=C0DF0TPMY). Though TBH, at the moment the Connect build being done on push is somewhat different from the build being done on a tag release, but that's another story.

File: rfd/0089-merge-webapps.md
Comment (ravicious) [inline rfd/0089-merge-webapps.md:37]: # How this will affect Connect

Generally not much as Connect is pretty self-contained. This is what comes to my mind on top of what Zac said.

## Default path to tsh

We'll have to update [the default path to tsh in dev mode](https://github.com/gravitational/webapps/blob/27c615b3ff6f317a85fac4aa28b8e73fa4aa0d28/packages/teleterm/src/mainProcess/runtimeSettings.ts#L18-L23). It's a simple change.

## Proto files

As mentioned by Zac, once webapps is merged into teleport, we want to streamline proto generation for Connect (#15955). Currently we update the protos in teleport, commit them there and then copy them over to webapps and commit the changes there as well. I suspect what we'll want to do is tell `buf` to generate Go proto files `lib/teleterm/api/protogen/golang` and the JS proto files inside `webapps/packages/teleterm/src/services/tshd/v1/`.

I'm not sure if #15955 needs to be implemented within this PR. It includes a lot of work already and we can live with copying the proto files for now.

File: rfd/0089-merge-webapps.md
Comment (ravicious) [inline rfd/0089-merge-webapps.md:57]: Does this mean that the merge of webapps into teleport will be backported to v8, v10 and v11?

File: rfd/0089-merge-webapps.md
Comment (ravicious) [reply rfd/0089-merge-webapps.md:37]: Oh sorry, #15955 is about an even more obscure issue with protos in Connect. What I had in mind were the tshd protos generated with `make grpc-teleterm`, [as described in Connect's readme](https://github.com/gravitational/webapps/tree/27c615b3ff6f317a85fac4aa28b8e73fa4aa0d28/packages/teleterm#generating-tshd-grpc-protobuf-files).

We don't have to worry about #15955 at all, nor about Connect protos in general. We'll just keep copying them over, the location will just change slightly. Once the merge is completed we can tweak `make grpc-teleterm` so that copying is not necessary. I'll make an issue about it as we already have a bunch of issues related to Connect's proto files.

File: rfd/0089-merge-webapps.md
Comment (zmb3) [reply rfd/0089-merge-webapps.md:130]: Or just `/web`.

File: rfd/0089-merge-webapps.md
Comment (zmb3) [reply rfd/0089-merge-webapps.md:135]: This unfortunately is not up for debate. We must enforce two reviewers for our SOC-2.

We got away with not enforcing 2 reviewers on webapps in the past because we required 2 approvals to update the submodule and incorporate the changes into teleport. Now that all the code will live in teleport, we'll be leveraging the same review process and requirements as any other teleport change.

File: rfd/0089-merge-webapps.md
Comment (zmb3) [reply rfd/0089-merge-webapps.md:37]: We should also add make targets for Teleport Connect that respect our existing OS/ARCH vars and correctly express the dependency on `build/tsh`.

File: rfd/0089-merge-webapps.md
Comment (ravicious) [reply rfd/0089-merge-webapps.md:135]: > This unfortunately is not up for debate. We must enforce two reviewers for our SOC-2.

Yes, I understand that. What I meant is not that we should maybe allow for just one reviewer but rather that in the past you had at least devs looking at your backend code and at least one fullstack looking at the frontend. Ending up with two backend devs looking at frontend changes or one backend and one fullstack dev feels just slightly worse in comparison.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:49]: Actually that sentence was poorly written, I've rewritten it :)

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:51]: I added a reference to using a `git worktree` with the rewrite of this paragraph.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:63]: This is mostly for those wanting to use `make full` to build a Teleport binary. As there will no longer be a `webassets` directory the UI will need to be built along with `make full` which requires node/yarn to do so.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:71]: Agreed, the second answer there is the approach that I've been playing with.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:135]: > Ending up with two backend devs looking at frontend changes or one backend and one fullstack dev feels just slightly worse in comparison.

We are able to add additional requirements to the reviewer script so that changes to the `webapps` folder use a different list of developers. I can expand this line item to include that script change as well.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:144]: What I meant to say was that it can just try to build the webassets not build and push. I'll update the copy.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:57]: Yes that's correct, well v9-11 :)

File: rfd/0089-merge-webapps.md
Comment (ravicious) [reply rfd/0089-merge-webapps.md:144]: Are we talking about the webapps-build CI job from `.cloudbuild/ci/build.yaml`? I don't think it pushes anything right now. 🤔

File: rfd/0089-merge-webapps.md
Comment (rudream) [inline rfd/0089-merge-webapps.md:111]: Why not retain the `/webassets` submodule and have a script that updates it on every UI PR merge? The same behaviour we currently have (where every `webapps` merge updates the `/webassets` in `teleport`), except within the same repo.

Since the majority of devs working in the `teleport` repo don't make any UI changes, it seems like an unnecessary waste of time to have to build the UI on every Teleport build, this also means they wouldn't be required to have node and yarn installed on their machines. Another approach to mitigate this could be to have a specific `make` script that builds the Teleport binaries but doesn't build the UI. Any thoughts?

File: rfd/0089-merge-webapps.md
Comment (codingllama) [inline rfd/0089-merge-webapps.md:128]: Any changes in these? Or are we keeping webapps build/lint checks contained in the webapps- jobs (which I would actually recommend)?

File: rfd/0089-merge-webapps.md
Comment (zmb3) [inline rfd/0089-merge-webapps.md:63]: Let's be explicit here and mention that a compiled version of the webassets will no longer be committed to source control.

File: rfd/0089-merge-webapps.md
Comment (zmb3) [inline rfd/0089-merge-webapps.md:56]:
```suggestion
of teleport, all supported release branches (v9, v10, v11) will have their respective
versions of webapps migrated into the `teleport` repository.
```

File: rfd/0089-merge-webapps.md
Comment (zmb3) [inline rfd/0089-merge-webapps.md:63]:
```suggestion
teleport binaries. A make target will be provided to perform a dockerized build for those who don't wish to install node and yarn on their workstation.
```

File: rfd/0089-merge-webapps.md
Comment (zmb3) [reply rfd/0089-merge-webapps.md:111]: My 2 cents:

I don't like that the source code and the webassets submodule within the same commit/tag can be out of sync.

Dev's won't need to build the UI on _every_ teleport build, they'll only need to build it once to have a local copy of the assets and then any time they want to recompile UI changes.

File: rfd/0089-merge-webapps.md
Comment (codingllama) [reply rfd/0089-merge-webapps.md:63]: +1, both for not needing node/yarn and for the determinism this brings.

File: rfd/0089-merge-webapps.md
Comment (zmb3) [inline rfd/0089-merge-webapps.md:39]: Today we have the ability to run `DEBUG=1 teleport start ...` in order to load webassets from disk instead of the ones embedded in the binary.

Let's not lose this capability. You'll have to make a change to Teleport (probably before the migration) to update where it looks for webassets. Please play around and see if you can get this working, then update the "Actions" section at the end of this document.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:128]: No changes to these, the webapps jobs will be moved over and run separately. This will allow us to skip out of jobs that don't need to run.... No need to run the `Teleport-Lint` job for example when changes were only made to the UI.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:130]: Updated the doc to use the `/web` folder name.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:39]: After playing with this a bit I think the best way to do this is to remove the `/webassets` submodule but keep the folder as the output location of the on-demand build. This will keep the functionality of `DEBUG=1` intact without any changes to `teleport` itself. I've updated the action below accordingly.

**Review by nklaassen — COMMENTED** [review]
Will all new github issues be opened in the teleport repo? Will existing issues be migrated? I'm glad to see that we plan to preserve git history

File: rfd/0089-merge-webapps.md
Comment (gzdunek) [inline rfd/0089-merge-webapps.md:126]: Somewhere here I'd add a point that Connect's build pipelines (in `teleport/dronegen`) needs to be updated - we won't have to clone `webapps` anymore.
> There's a number of // TODO in dronegen that will also be removed since the teleport connect code will already be present by default 

(from Zac's comment)

Comment (hatched) [reply rfd/0089-merge-webapps.md:126]: done!

**Comment by hatched** [comment]
@nklaassen Yeah Under the webapps repository section I outline that we will transfer all of the existing webapps issues into the `teleport` repo and add the already existing `ui` label.

**Review by kimlisa — APPROVED** [review]

File: rfd/0089-merge-webapps.md
Comment (ravicious) [inline rfd/0089-merge-webapps.md:91]: How do we plan to orchestrate this? We cannot tell people to stop working until the switch is performed. But the way these points are stated here doesn't seem to imply that it'll be a seamless process.

Assuming the git history is kept, I think it could be possible to move any WIP branches by adding a remote in your local webapps clone which points to teleport, then push the branch to that remote, clone the branch in your local clone of teleport and rebase the branch on top of already migrated master of teleport.

File: rfd/0089-merge-webapps.md
Comment (ravicious) [inline rfd/0089-merge-webapps.md:140]: How do we feel about sticking to the current process of manually picking reviewers?

It's not something we've discussed within the frontend team and as I mentioned [in my earlier comment](https://github.com/gravitational/teleport/pull/16413#discussion_r975315309), I feel it might be disruptive for the day-to-day work.

File: rfd/0089-merge-webapps.md
Comment (ravicious) [reply rfd/0089-merge-webapps.md]: Ping; I don't think this jobs pushes webassets?

File: rfd/0089-merge-webapps.md
Comment (codingllama) [reply rfd/0089-merge-webapps.md:67]: > Does anyone actually develop the teleport binary without embedding webassets?

I run debug sessions with `DEBUG=1 DEBUG_ASSETS_PATH=/path/to/webassets`. Super helpful to figure out flows that you are not familiar with.

Comment (ravicious) [reply rfd/0089-merge-webapps.md:67]: > I run debug sessions with `DEBUG=1 DEBUG_ASSETS_PATH=/path/to/webassets`. Super helpful to figure out flows that you are not familiar with.

Could you give an example of a scenario where you'd use this? It's not clear to me why I'd want to use those two env vars over recompiling teleport with embedded webassets or proxy updated Web UI to an existing cluster with `yarn start-teleport --target=https://example.com:3080/web`.

I suppose `DEBUG=1 DEBUG_ASSETS_PATH=…` replaces `yarn start-teleport` if you don't have Node.js tooling installed? Because then you can go to webapps, execute `make build-teleport` and then point `DEBUG_ASSETS_PATH` at assets that were just generated for you through Docker.

Comment (codingllama) [reply rfd/0089-merge-webapps.md:67]: That is not for Web UI development, but part of the invocation I use to debug the `teleport` binary itself for backend dev. I generally don't care about the webassets version, I just need the server to start.

**Comment by hatched** [comment]
@ravicious good point about needing to make commits to older releases. It looks like we update v8 often enough that archiving the webapps repository now would cause undue friction in updating those older releases. I'll update the references about archiving webapps to instead update the readme to point developers to the Teleport repo and we can follow-up in 6mo to see if we can safely archive the repo at that time.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:67]: I could potentially see a env variable `NO_UI=1 make full` or something to skip making the UI but I think that's out of scope for now.

[this comment](https://github.com/gravitational/teleport/pull/16413#discussion_r1001123524) outlines how this will work a bit. So one of the steps in the build process will create the built webassets and then the rest of the build will carry-on as it does now. I feel like this will be the quickest way to get this integrated and then we can adjust the process afterwards if we feel it needs adjusting.

File: rfd/0089-merge-webapps.md
Comment (hatched) [reply rfd/0089-merge-webapps.md:91]: Ultimately the code will be 'moved over' one day so all work can continue on the existing branches until that time. The issue that this is trying to solve is that there are a number of open PR's that are stale so in having a cleaner list of open PR's will make the move simpler.

File: rfd/0089-merge-webapps.md
Comment (ravicious) [reply rfd/0089-merge-webapps.md:67]: > That is not for Web UI development, but part of the invocation I use to debug the `teleport` binary itself for backend dev. I generally don't care about the webassets version, I just need the server to start.

@codingllama Ah, so you compile `teleport` without webassets and then use DEBUG & DEBUG_ASSETS_PATH just so that you can start the server?

> (…) one of the steps in the build process will create the built webassets and then the rest of the build will carry-on as it does now. I feel like this will be the quickest way to get this integrated and then we can adjust the process afterwards if we feel it needs adjusting.

@hatched Yeah, it'll definitely be the quickest way. And if other people use a similar workflow for backend dev to what Alan described then it might not be as disruptive.

File: rfd/0089-merge-webapps.md
Comment (ravicious) [reply rfd/0089-merge-webapps.md:92]: > Ultimately the code will be 'moved over' one day so all work can continue on the existing branches until that time. The issue that this is trying to solve is that there are a number of open PR's that are stale so in having a cleaner list of open PR's will make the move simpler.

Ah, thanks for the explanation, that makes sense. It looks like we were on the same page but I think this point about people being able to continue work on their regular branches wasn't expressed in the RFD so I wasn't sure how you planned around this.

File: rfd/0089-merge-webapps.md
Comment (ravicious) [reply rfd/0089-merge-webapps.md]: It's still not clear to me what kind of pushing this job does and will not do in the future. Am I mistaking it with some other job?

**Review by ravicious — APPROVED** [review]
I don't want this to sound overly dramatic but this RFD includes less details than I'd initially assumed it'd include.

I expected it to go into the nitty-gritty of preserving the history and how the dev process is going to be affected. But perhaps I misunderstood the goal behind the RFD which was more to sketch a general plan which will enable you to move forward rather than having everything planned to the last detail (which probably would be impossible to do with a change of this scope anyway).

But yeah, overall the plan looks fine I think!

Could you address my last question about webapps-build before merging? That's the only remaining thing that I don't fully understand.

---

Let us know if you run into problems regarding Connect's build process on drone because I feel like there's a lot of siloed knowledge which will take 10 minutes to explain vs two hours of trying to understand it by reading dronegen scripts.

**Review by gzdunek — APPROVED** [review]

**Comment by hatched** [comment]
@ravicious Thanks for voicing your concern about the level of detail here. You're right that this is to just be a more higher level overview of the necessary steps. There a number of changes that need to occur to make this transition seamless (or as seamless as it can be) and have them all be detailed here. I also expect that there to be some unforeseen consequences that'll need to be tackled....an "unknown unknown" as it were :).

I'll fix the description of the `webapps-build` job as all it does is check to see if it successfully builds the project. I think this was a copy/paste error.

**Review by nklaassen — COMMENTED** [review]
We are already reaching the breaking point for how many PRs we can merge to teleport per day with our current CI processes. My biggest concern here is that adding all webapps PRs to the teleport repo may push us into a state where we are writing new PRs faster than we can merge them.

I really hope that we can get a proper merge queue before we complete this effort. If we plan to merge these repos without a merge queue, I'd like to see some numbers on `(PRs per day) * (CI time + Update Branch latency)`.

Can you add a section saying either "we won't merge the repos until we have a merge queue" or "this will make it take X hours longer to merge all otherwise ready PRs per day and we think that's acceptable"

Other than that, I think it will be a big improvement as long as we can keep our CI under control

**Comment by hatched** [comment]
@nklaassen in the past 30 days there were approximately 75 PR's to the webapps project. 40 of which were backports off `master` so there are really only 34 additional, or ~1.5 per day, prs that will be of consequence to the primary development flow on the teleport repo. The `build` job takes less than 6m on average and the `test` job, 11m.

In the RFD I have included comments about optimizing the jobs so they only run when necessary. So for example, we wouldn't be running the UI tests when only backend changes were made. Outside of scope for this RFD however there are also some low hanging fruits on the CI jobs in teleport we can optimize, Like not running `teleport-lint` (10m) or `teleport-oscompatibility` (10m) for docs changes.  `codeql` is also taking 25m to build teleport.

This will mean that when a frontend PR lands it'll delay the landing of a backend PR by 30minutes on average but this is primarily a symptom of the backend CI jobs than an abundance of frontend PRs. Because of this I don't think that we need to delay it until the merge queue has been implemented.

**Comment by nklaassen** [comment]
@hatched thanks. There's an average of ~1.5 per day, but the bigger concern is actually the peak number of PRs we try to merge right before a release, when it becomes very difficult to merge things even now. But I hear that the GHA is "Coming Soon" so hopefully we can start using the merge queue before the next major release.

It's great for frontend devs if they can merge in 11 minutes, but if a backend PR is 10 minutes into its CI run and a frontend PR starts, completes, and merges in 11 minutes, invalidating the backend PR, that is worse from the backend dev's perspective :) Obviously not the frontend team's fault and the real solution is a merge queue or similar.

With an average of +1.5 PRs/day I don't think this will be disastrous, could you just add a short section to the doc briefly summarizing the expected impact and I will approve?

**Review by nklaassen — APPROVED** [review]

**Comment by hatched** [comment]
Thanks everyone for the comments, opinions and reviews!

**Pull request merged**
